### PR TITLE
feat: Auto teleport V1 

### LINKED ISSUE #7643: Auto teleport V1
- this issue got pretty messy, thus creating new one. 
https://github.com/kodadot/nft-gallery/issues/4917

- here we go. After some iterations, simplifications, and an almost complete modal transition, here is our auto teleport. 

# Overview

- **Auto Teleport**
  - Automates fund transfers between chains where the users have the most assets to a chain where they are making the action. 
  - Can be activated or deactivated via a toggle button.

- **First-time User Onboarding Modal**
  - Activated when Auto Teleport is switched on for the first time.
  - Highlights benefits: 
    - *Effortless Transfers*
    - *Time-saving*
    - *Smooth Experience*
  - Informs user to approve an extra transaction each time the feature is needed.
  - then button: "Got it!"

- **Disabled State of Auto Teleport**
  - When no funds are available on any chain, display "Not Available" label with a tooltip icon. 

- **Signing Modal**
  - Contains steps for approving transactions.
  - Includes 'Checking Funds' step to confirm successful transfer of required funds before proceeding.

- **Signing Modal**
  - there is also a tooltip for the auto teleport in the transaction summary - which shows users the path we chose for them, introduction, option to learn more. 

------

# Design
- updated in figma handoff: https://www.figma.com/file/JXNtryaNYhY0XJNArzvPQx/Handoff?type=design&node-id=4354%3A15507&mode=design&t=b1QX6PAfX3d9IHP1-1
image

## Auto teleport row

- **Tooltip**: Displayed for users to explain the functionality and working of Auto Teleport when hovered or clicked. Correspondingly changes based on user fund availability across chains.

- **Auto Teleport Modal States**:
  - labeled in the designs
  - **Sufficient Funds**: The Auto Teleport feature is not displayed when all funds are sufficient for a transaction.
  - **Insufficient Source Chain Funds**: When funds on source chain are insufficient but available on other chains, the feature row is displayed allowing teleport and purchase confirmation with Auto Teleport turned on. If turned off, it provides low-fund warning.
  - **No Funds**: When no enough funds are available at all, the row is disabled and shows "Not Available" text.

image

## Tooltip 

- **Tooltip when no funds are available**: Explains the Auto Teleport feature and tells users why it's currently unable to function due to no available funds.
image

- **Tooltip when funds are available on other chains**: Details how Auto Teleport can transfer assets from the user's most asset-rich chain to enable the transaction.
image

## Onboarding Modal (First-time user)
- Displayed for first-time activation of Auto Teleport. (from off -> on) 
- displayed over the current modal, it does not replace, its on top of it. You can close it both ways, got it button or x-mark 
image

## Signing 
- similar to #7552 

- **Header**: The modal has a clear heading, "Sign Transactions"

- **Loader wheel**: A loader or progress indicator is displayed next to each transaction section (Teleporting/Bridging, Checking Funds, and Buying NFT). This loader becomes active when the respective transaction is ongoing, providing visuals about the current process status.

- **Transaction sections**:. These include:
   1. Teleporting/Bridging
   2. Checking Funds
   3. Buying NFT

- **Transaction state labels**: Each transaction have labels showing transaction states: 'please sign' (for initial user approval), 'transaction pending', and 'completed' (once all operations within that step are successfully executed) It has also link to a transaction which when clicked will open new window on transaction explorer  

image

states for the single row: 
image

### End
- one thing missing is the confirmation screen - when everything is okay. I will make this later. 
- again, if you have any questions, let me know! Thank you

### LINKED ISSUE #7718: xmark modals
It's a really small detail, but now the x-mark is not aligned on the x-axis. 
image

_Originally posted by @exezbcz in https://github.com/kodadot/nft-gallery/issues/7660#issuecomment-1765153374_

### PR DESCRIPTION
**Pull request by hassnian**
**Thank you for your contribution** to the [KodaDot - One Stop Shop for Polkadot NFTs](https://kodadot.xyz).

👇 __ Let's make a quick check before the contribution.

## PR Type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

## Context

⚠️WIP ⚠️ 

- [x] Closes #7643
- [x] Closes #7718

#### Did your issue had any of the "$" label on it?

- [x] Fill up your DOT address: [Payout](https://kodadot.xyz/dot/transfer?target=13QUj3pZyFNfYj4AM336hRdyLQbevj5H3sR4PKmLEXLdwZhh)

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [x] My fix has changed **something** on UI; 

CleanShot 2023-10-26 at 08 35 18@2x

CleanShot 2023-10-30 at 10 26 31@2x

### <samp>🤖 Generated by Copilot at c55e671</samp>

This pull request improves the functionality and user experience of the NFT gallery by adding support for cross-chain transfers of NFTs and displaying multiple balances for different chains and tokens. It introduces new composables, components, and utils to handle the logic and UI for auto teleport, multi balance, and teleport routes. It also simplifies and refactors some existing code in `MultipleBalances.vue` and `Teleport.vue`.

### <samp>🤖 Generated by Copilot at c55e671</samp>

> _We are the masters of the chains, we teleport with ease_
> _We balance all the assets, we use the composables_
> _We create the components, we improve the UI_
> _We are the kodadot rockers, we code with `Vue` and `TS`_

### CODE REVIEW

**Comment by exezbcz** [comment]
ready for first round 👀

**Comment by exezbcz** [comment]
> or only needed

only needed

> almost done, had a busy week with something else , I expect to finish this weekend and have it ready by monday, want to refactor some things and just need to added it in all the other modals

nice! 

> so this should be added for every modal where you spend tokens , transfer , offer , buy , list (?), right, etc ?

yes, its mainly for actions no assethubs

**Comment by exezbcz** [comment]
actually bought nft through the autoteleport, will post some feedback today/tomorrow

looking really nice sir

**Comment by hassnian** [comment]
@exezbcz 

for transfers show the auto teleport be here , ie auto teleport -> bridge -> funds -> confirm trasnfer (step 3) ? and then -> show confirm modal 

kinda makes no sense right ? not sure 

CleanShot 2023-10-28 at 17 03 05@2x
or inside the confirm modal ?

**Comment by hassnian** [comment]
atm bsx seems to be down I can't add the auto teleport to the offer , will do as soon as possible , also I've noticed some small issues, that will fix ASAP 

but I think its time for a first round 

@kodadot/qa-guild @exezbcz 👀

@roiLeo one thing that I noticed is that I ended up having all the logic of the auto teleport inside the button itself , I could have the modals outside but seems to me that the same code will be duplicated, not sure , will be waiting for your review :D 


thanks guys

**Comment by exezbcz** [comment]
# Feedback 

here we go

# First 
- I think it will be nice to make the auto teleport turned on by default
- after you enter the confirmation modal
image
- the blurred shape is bit off - could you check figma please
image
image
- this icon to grey

# Onboard screen 
- good, but the icons did not load for me
telegram-cloud-photo-size-4-5902147666391974073-x

- the main heading should be 25px bold text 


# Signing 
image
- remove the "networks" from quotes, that was intended as placeholder

image
- place the icon close and aligned with the text - as in designs 

image
- grey loading circles please 

- it also does not detect the state correctly - I haven't signed anything and its already in the pending state 
image

image
- buy nft button active before the green checks are displayed 

# buy nft 
image
- I will put this step automatically for signing, so user does not have to click the button twice. So, from checking funds it will automatically go to buying nft step 

image
- this state is not correct
- the subheading should be visible after you are on the step - not before
- before that, the subheading is not there


again, the transaction is pending even without me signing the transaction. 
image

this is nice, good job!
image

# done 
image
- one thing i missed in the designs is to add view tx hyperlink to the buying nft transaction too

# ad modal and some styles 
image
- the top part should be fixed, and the rest scrolls, as it is with the mass listing

**subheadings**: 
image
these subheadings to 12px font, please

-------------

# End
once again, really nice job! This is all I found in a pretty big issue. We will do some follow-ups, too.

**Comment by prury** [comment]
that's a lot of work @hassnian! congrats! 

playwright detection(when the user has no funds on a chain): 
image

**Comment by prury** [comment]
image

**Comment by hassnian** [comment]
missing , create nft and collections , will do asap

**Comment by exezbcz** [comment]
> offer -> has old flow of slider button


- https://github.com/kodadot/nft-gallery/issues/6476

**Comment by exezbcz** [comment]
> nft creation on rmrk seems to have an extra step what should happen here ?

in the signing modal or? 

if there is additional step then yes

**Comment by prury** [comment]
Attempt at buying RMRK2 NFT with funds being teleported from Kusama Asset Hub

**Comment by hassnian** [comment]
> Attempt at buying RMRK2 NFT with funds being teleported from Kusama Asset Hub (was not able to):
> 
>  Rmrk2.with.funds.from.Assethub.mp4 
>  second.attempt.mp4

seems like it couldn't connect with bsx, bsx seems to be down a lot

File: components/common/autoTeleport/AutoTeleportActionButton.vue
Comment (roiLeo) [reply components/common/autoTeleport/AutoTeleportActionButton.vue:53]: could be replaced by a `div`

File: components/common/autoTeleport/AutoTeleportActionButton.vue
Comment (roiLeo) [inline components/common/autoTeleport/AutoTeleportActionButton.vue:56]: I would change it with a text button (to avoid undefined href)

File: components/common/autoTeleport/AutoTeleportActionButton.vue
Comment (roiLeo) [inline components/common/autoTeleport/AutoTeleportActionButton.vue:268]: maybe use `size="large"` prop

File: components/common/autoTeleport/AutoTeleportModal.vue
Comment (roiLeo) [inline components/common/autoTeleport/AutoTeleportModal.vue:122]: ts warn here

File: components/common/autoTeleport/AutoTeleportModal.vue
Comment (roiLeo) [inline components/common/autoTeleport/AutoTeleportModal.vue:168]: ts warn

File: components/common/autoTeleport/AutoTeleportModal.vue
Comment (roiLeo) [inline components/common/autoTeleport/AutoTeleportModal.vue:225]: how's mobile?

File: components/common/autoTeleport/AutoTeleportTooltip.vue
Comment (roiLeo) [inline components/common/autoTeleport/AutoTeleportTooltip.vue:55]:
```suggestion
          >{{ $t('helper.learnMore') }}</a
```

File: components/common/autoTeleport/AutoTeleportTooltip.vue
Comment (roiLeo) [inline components/common/autoTeleport/AutoTeleportTooltip.vue:51]: link to?

File: components/common/autoTeleport/AutoTeleportWelcomeModal.vue
Comment (roiLeo) [inline components/common/autoTeleport/AutoTeleportWelcomeModal.vue:53]: link to?

**Comment by hassnian** [comment]
@prury give me a day or two i've had to change some stuff to accommodate for nft and collection creation and you might end up retesting buy and list , i'll ping you giving you the  green light

**Comment by prury** [comment]
> > Attempt at buying RMRK2 NFT with funds being teleported from Kusama Asset Hub (was not able to):
> > Rmrk2.with.funds.from.Assethub.mp4
> > second.attempt.mp4
> 
> seems like it couldn't connect with bsx, bsx seems to be down a lot

yes...ugh, haha

**Comment by prury** [comment]
> @prury give me a day or two i've had to change some stuff to accommodate for nft and collection creation and you might end up retesting buy and list , i'll ping you giving you the green light

take your time my dude! allright

**Comment by hassnian** [comment]
@prury  feel free to test , list , buy , mint nft , mint collection

**Comment by yangwao** [comment]
Is it ready for review & merge? 
As it's quite big, appreciate that @hassnian and quite long out there 😄
image
image

**Comment by hassnian** [comment]
> Is it ready for review & merge? 
> 
> As it's quite big, appreciate that @hassnian and quite long out there 😄
> 
> image
> 
> image
> 
> 

yes, as far as i can tell

**Comment by prury** [comment]
Some tests: 
Buy now on RMRK1, having almost 0 balance on Kusama, and $3 on Ahk: ❌ 
https://github.com/kodadot/nft-gallery/assets/36627808/6b787e69-b4d6-47e9-b4be-94a0f6ce80c2

Listing on RMRK2 having almost 0 balance on Kusama, and $2 on Ahk: ❌ 

Create NFT on RMRK2, having almost 0 balance on Kusama and $3 on Ahk: ❌ 
https://github.com/kodadot/nft-gallery/assets/36627808/75799eaa-b5aa-44a5-a23a-30e9f9e2ba04

Buy now on Assethub: ✔️ 
https://github.com/kodadot/nft-gallery/assets/36627808/1ee75f24-92da-4fd7-9a7d-ab40cf7a6631

Listing on Ahk: ❌ (reduced my assethub balance to: image) - teleport occurs but listing does not
https://github.com/kodadot/nft-gallery/assets/36627808/f136c701-1614-495f-9479-0c7bf5775cb3

Observations:
- Balances on sidebar take a lot more longer to load in this deployed version if compared to canary

File: components/common/autoTeleport/AutoTeleportWelcomeModal.vue
Comment (hassnian) [reply components/common/autoTeleport/AutoTeleportWelcomeModal.vue:53]: not available 
conversation with exezbcz 
CleanShot 2023-11-07 at 09 41 15@2x

File: components/common/autoTeleport/AutoTeleportModal.vue
Comment (hassnian) [reply components/common/autoTeleport/AutoTeleportModal.vue:248]: I think this is recurring issue, confirm buttons are a size that is not the same as size large and also changes the button font size  to a larger one 


we might want to do something since it's spreading ...
CleanShot 2023-11-07 at 09 46 29@2x

File: components/common/autoTeleport/AutoTeleportActionButton.vue
Comment (hassnian) [reply components/common/autoTeleport/AutoTeleportActionButton.vue:56]: fixed

File: components/common/autoTeleport/AutoTeleportModal.vue
Comment (hassnian) [reply components/common/autoTeleport/AutoTeleportModal.vue:245]: fixed

**Comment by hassnian** [comment]
@prury try now thanks

File: components/common/autoTeleport/AutoTeleportModal.vue
Comment (roiLeo) [reply components/common/autoTeleport/AutoTeleportModal.vue:249]: oh no ... 😦 
let's open

File: components/common/autoTeleport/AutoTeleportWelcomeModal.vue
Comment (roiLeo) [reply components/common/autoTeleport/AutoTeleportWelcomeModal.vue:53]: alright if still not provided you should hide it (`.is-hidden`)

File: components/common/autoTeleport/AutoTeleportModal.vue
Comment (roiLeo) [reply components/common/autoTeleport/AutoTeleportModal.vue:249]: continue #7957

**Comment by prury** [comment]
> @prury try now thanks

will be doing a properly analysis on my morning, got caught in some stuff today

File: components/common/autoTeleport/AutoTeleportModal.vue
Comment (Jarsen136) [inline components/common/autoTeleport/AutoTeleportModal.vue:233]: why use setTimeout

**Comment by yangwao** [comment]
wow, I noticed this, congratulations @hassnian !
waiting for approval
image

**Comment by prury** [comment]
https://github.com/kodadot/nft-gallery/pull/7711#issuecomment-1796963324 updated

@yangwao if you want me to test auto-teleport on Polkadot Asset Hub, please fund the following account for collection and nft creation:  1CBmjqJvV1ha8BPQ5JAXGEzeJBLmGHNG9gYg96eC1dnjH5W

File: components/common/autoTeleport/AutoTeleportModal.vue
Comment (hassnian) [reply components/common/autoTeleport/AutoTeleportModal.vue:233]: for nft creation there is a redirect to the nft page, closing immedialty didn't feel quite right so that's why I added the auto close with a delay

File: composables/autoTeleport/useAutoTeleportTransition.ts
Comment (hassnian) [reply composables/autoTeleport/useAutoTeleportTransition.ts:8]: ups autoimport

File: components/buy/Buy.vue
Comment (daiagi) [reply components/buy/Buy.vue:77]: conversation is closed but the suggestion is not committed.
why?

Comment (hassnian) [reply components/buy/Buy.vue]: I was making some changes and had not pushed the changes with the fix

Comment (daiagi) [reply components/buy/Buy.vue]: OK i seen you have committed the suggestion
please as a good practice to help us all track what is done and what is not
allow the reviewers to close the conversation they started

File: components/buy/Buy.vue
Comment (daiagi) [reply components/buy/Buy.vue]: here again.
do not close conversations on behalf of the reviewer

File: components/common/autoTeleport/AutoTeleportModal.vue
Comment (daiagi) [reply components/common/autoTeleport/AutoTeleportModal.vue:233]: if so then move it to some const with a good name please
avoid magic numbers

File: utils/balance.ts
Comment (daiagi) [inline utils/balance.ts:74]: mind you there is no need for `return await`
a simple `return` would return a promise the same

**Review by daiagi — APPROVED** [review]
nice work!

1 thing is:
this link should point to the transaction, currently, it points to the block
image

File: components/buy/Buy.vue
Comment (hassnian) [reply components/buy/Buy.vue]: duly noted

File: utils/balance.ts
Comment (hassnian) [reply utils/balance.ts:74]: indeed, only first one is needed

File: components/buy/Buy.vue
Comment (hassnian) [reply components/buy/Buy.vue]: noted

File: components/common/autoTeleport/AutoTeleportModal.vue
Comment (hassnian) [reply components/common/autoTeleport/AutoTeleportModal.vue:233]: done

File: components/buy/Buy.vue
Comment (Jarsen136) [reply components/buy/Buy.vue]: So is there any clarification here? 👀

**Review by Jarsen136 — APPROVED** [review]
Please resolve the code conflict

**Comment by hassnian** [comment]
@prury 

- listing should be fixed
-  collection creation 

> Second try: ❌
> https://github.com/kodadot/nft-gallery/assets/36627808/1f0a476c-3545-4325-91f9-92bcf9c4843f
> https://assethub-kusama.subscan.io/tx/0x19aac0a9c95e1af819ad9e5d6ef468d923c2c6eaf3000ca07afc4541c243266f

   i get the same error on canary  not sure why , any idea @kodadot/internal (check video 👆), 

already checked  txFee and  collectionTotalDeposit 


> 1 thing is: this link should point to the transaction, currently, it points to the block image

can do that on follow up or when I'm back 

fyi, got something to do and will be afk for a couple of days , will be  back on Monday maybe earlier

**Comment by yangwao** [comment]
> can do that on follow up or when I'm back

yes, I'm gonna test it myself and would be pro merging it as I sense some dragons might appear or some edge cases over time.

**Comment by yangwao** [comment]
Noticed a few errors on the sidelines, but probably unrelated, as everything has worked for me

image

image

image

**Comment by yangwao** [comment]
lot of effort and worked perfectly, looking forward!
pay 230 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $230 @ 5.13 USD/DOT ~ 44.834 $DOT
🧗 [13QUj3pZyFNfYj4AM336hRdyLQbevj5H3sR4PKmLEXLdwZhh](https://polkadot.subscan.io/account/13QUj3pZyFNfYj4AM336hRdyLQbevj5H3sR4PKmLEXLdwZhh)
🔗 [0x10eb2d33b999f723c67a57c261bfda92098f2d043d01f91dfce555d6436531c9](https://polkadot.subscan.io/extrinsic/0x10eb2d33b999f723c67a57c261bfda92098f2d043d01f91dfce555d6436531c9)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Comment by yangwao** [comment]
@prury if there is something, let's create follow up, for me buying worked and that was primary goal for this one, happy to learn if there any bugs in follow up isues

**Pull request merged**
